File: Cpp/sources/quick.cpp
#include <iostream>
#include <ctime>
#include "../headers/swap.h"
#include "../headers/value.h"

using namespace std;

int partition(int* a, int l, int r)
{
    int p = a[l+(r-l)/2];

    while (l <= r)
    {
        while (a[l] < p)
            ++l;
        while (a[r] > p)
            --r;
        if (l <= r)
        {
            swap(a, l, r);
            ++l;
            --r;
        }
    }
    return l;
}

void quick_sort(int* a, int l, int r)
{
    int pl = partition(a, l, r);

    if (l < pl-1)
        quick_sort(a, l, pl-1);
    if (r > pl)
        quick_sort(a, pl, r);
}

int main()
{
    int s = 1000000; // Changeable size
    int *a = rev_pn(s);

    clock_t start = clock();

    quick_sort(a, 0, s-1);

    clock_t ended = clock();
    double total = (double)(ended - start) / CLOCKS_PER_SEC;

//    for (int i=0;i<s;++i)
//        cout << a[i] << " " << endl;
    cout << total << " seconds";

    free(a);
    return 0;
}

File: sources/quick.py
import time
from Python.others.swap import *
from Python.others.value import *


def partition(a, l, r):
    p = a[l+(r-l)//2]

    while l <= r:
        while a[l] < p:
            l+=1
        while a[r] > p:
            r-=1
        if l <= r:
            swap(a, l, r)
            l+=1
            r-=1
    return l


def quick_sort(a, l, r):
    pl = partition(a, l, r)

    if l < pl-1:
        quick_sort(a, l, pl-1)
    if r > pl:
        quick_sort(a, pl, r)


# Changeable size
s = 1000000
a = rev_pn(s)

start = time.time()

quick_sort(a, 0, len(a)-1)

end = time.time() - start

print(a)
print(end, "seconds")
